### ISSUE TITLE
transferred from fastify-autoload, v3.13.0 as tag missing, general issue with renamed packages and missing tags

### ISSUE THREAD

**Issue by DanielRuf** [fastify/help#929]
### Prerequisites

- [X] I have written a descriptive issue title
- [X] I have searched existing issues to ensure the issue has not already been raised

### Issue

Can you please push the tag for v3.13.0 for https://github.com/fastify/fastify-autoload/releases? Maybe other tags are also missing.

https://github.com/fastify/fastify-autoload/issues/234 was closed without any comment.

**Comment by Uzlopak** [comment]
Transferred. Should be discussed if this is is something we want to tackle.

**Comment by metcoder95** [comment]
The issue is that a missed `v3.13.0` tag is causing ORT to fail as the repository does not contain the commit hash for the given tag, as it shows an inconsistency with the repository history and their releases. Did I understood correctly?

**Comment by Uzlopak** [comment]
If this needs to be solved we should maybe create another branch independent from the default branch, like described here
https://link-intersystems.com/blog/2014/11/23/how-to-create-an-independent-branch-in-git/

Than we download the npm content, and push it as a commit to this branch. Tag it as the missing one.

**Comment by Eomm** [comment]
Some ref:

- https://github.com/fastify/fastify-autoload/pull/233
- https://github.com/fastify/deprecate-modules

TLDR: the tool created:
- 1 pr with the new name
- 1 local repo with the scaffolding code to deprecate the old names: https://www.npmjs.com/package/fastify-autoload?activeTab=code

I think to solve this we would need to:

- download the repo and create a `vX` branch and commit the scaffolding
- create a GH release with an older date
- npm should show the deprecation message (it is not https://www.npmjs.com/package/fastify-autoload?activeTab=readme)

We should do it 96 times 😄 https://github.com/fastify/deprecate-modules/blob/master/lib/modules.json

> If there is no actual commit and tag, then we have problems with the provenance of this package.

Don't use that version.. what's the problem?

I think the only thing to do here is to run a `npm deprecate <package-spec> <message>` against those 96 versions.

**Comment by DanielRuf** [comment]
> npm should show the deprecation message (it is not https://www.npmjs.com/package/fastify-autoload?activeTab=readme)

if you mean the "red" banner in deprecated packages, you have to manually mark this as deprecated via npm CLI or on npmjs.com by doing so

https://docs.npmjs.com/cli/v9/commands/npm-deprecate?v=true

https://docs.npmjs.com/deprecating-and-undeprecating-packages-or-package-versions

**Comment by DanielRuf** [comment]
> Don't use that version.. what's the problem?

We can not yet upgrade / migrate to the new packages for specific reasons that have to be discussed with the customer. So we can only update to the latest version which is helpful to have the deprecation notices for the future upgrades.

But using the documented way to deprecate packages and versions described in my previous comment is the better solution imho.

**Comment by DanielRuf** [comment]
> I think the only thing to do here is to run a `npm deprecate <package-spec> <message>` against those 96 versions.

That makes at least much more sense than these untagged / unreleased package versions (missing in the GitHub repo but on npmjs). With the upcoming Cyber Resilience Act the topic "provenance" will be more important than ever. Especially for correct SBOMs (Sofware Bill of Material)

See also https://docs.npmjs.com/generating-provenance-statements for the future.

**Comment by jsumners** [comment]
As noted above, there is not a commit to tag. The lengthy discussion in which the plan was thoroughly detailed is also linked above.

I recommend pinning your packages to the prior version if you are unwilling/unable to migratd to the currently supported versions.

I do not think it is worth the effort to go back through all of the repos and try to rewrite some sort of new history in order to satisfy some auditing tool. I do think it would be a good idea to issue an `npm deprecate` for all of the unsupported packages; it's just that no one has taken the time to do so.

> With the upcoming Cyber Resilience Act the topic "provenance" will be more important than ever.

The software in question is community developed and supported open source software. It is clearly labeled "use at your own risk." Users of such software are responsible for following whatever applicable laws are present in their region.

**Comment by DanielRuf** [comment]
> I do think it would be a good idea to issue an npm deprecate for all of the unsupported packages; 

Agreed.

> The software in question is community developed and supported open source software. It is clearly labeled "use at your own risk." Users of such software are responsible for following whatever applicable laws are present in their region.

It seems there is some misconception, probably caused by the news around this law. You can ignore the details of the law, SBOMs and code provenance are often required things in the (enterprise) software world and are becoming more important to reduce cases of supply chain attacks.

See it as a simple heads up, all future npmjs releases should have a Git tag in the linked GitHub repo. That's all.

I am totally aware of the fact that the past can not be changed and that the effort is too much to do it for past releases.

**Comment by jsumners** [comment]
> SBOMs and code provenance are often required things in the (enterprise) software world and are becoming more important to reduce cases of supply chain attacks.

The "use at your own risk" caveat still applies. "Enterprise" and any other organizations/individuals are responsible for doing their own due diligence on the software they use.


> I am totally aware of the fact that the past can not be changed and that the effort is too much to do it for past releases.

👍

**Comment by Uzlopak** [comment]
There is no rule that says, that we need to use git for npm packages. Heck it is not even a rule to refer to a repository in package.json! There is no rule that says, that we need to use a source code repository system which supports a functionality similar to tags. 

I actually looked up in the Cyber Resilience Act and there is nothing about provenance. 

And I kind of doubt, that we will face a sanction by von der Leyen, because we are missing a corresponding git tag for a npm package, which is not getting updates anymore. :D

Keep in mind: 

All maintained npm packages of format @fastify/x are actively maintained and should have for every released version on npm a corresponding git tag. So ORT should not report any problems with the actively maintained packages. 

If somebody uses ORT and gets the report, that in fastify-autoload a npm version is not having a corresponding git tag, than this person should reconsider the priorities, as using fastify-autoload and not @fastify/autoload indicates the use of old software and should invest time and material to upgrade his own code and use new packages.

**Comment by DanielRuf** [comment]
> There is no rule that says, that we need to use git for npm packages. Heck it is not even a rule to refer to a repository in package.json! There is no rule that says, that we need to use a source code repository system which supports a functionality similar to tags.

I think there is a misunderstanding now. I never said that. All other releases of you have proper Git tags. It was just forgotten to do for the (not so optimal) process of deprecation.

All I'm asking now (see my latest comments) and that I agree is: the old packages should be properly deprecated.

I can accept that the tags are missing and that can not be changed now.

> I actually looked up in the Cyber Resilience Act and there is nothing about provenance.

Take a look at how SBOMs work and what ORT does. But I don't want to get into detail now. It makes not much sense to discuss the CRA now. It was just one more example why releases should have a proper Git tag. Not more.

**Comment by DanielRuf** [comment]
> as using fastify-autoload and not @fastify/autoload indicates the use of old software and should invest time and material to upgrade his own code and use new packages.

Again, all I'm asking at the moment is to do the `npm deprecate ...` step. The rest is fine for me since the original discussion was linked and you all provided more details. And I agree with all of you, the past can not be changed. For the future, deprecations should be probably differently handled. I think we can all agree on that point.

**Comment by mcollina** [comment]
@jsumners do you still have the list of the modules we renamed?

**Comment by DanielRuf** [comment]
@mcollina I guess the 96 name entries at https://github.com/fastify/deprecate-modules/blob/master/lib/modules.json are meant.

**Comment by jsumners** [comment]
> @jsumners do you still have the list of the modules we renamed?

<details>
<summary>I got you</summary>

![giphy](https://github.com/fastify/help/assets/321201/6fb53d1a-df92-4e2b-90c9-686e97cd7c32)


</details>

+ https://github.com/fastify/fastify/issues/3733#issuecomment-1111195661
+ https://github.com/fastify/fastify/issues/3733#issuecomment-1111188683

**Comment by Eomm** [comment]
> We can not yet upgrade / migrate to the new packages for specific reasons

For the logs: the next major version contains only the renaming stuff. It is safe to bump
https://github.com/fastify/fastify-autoload/releases/tag/v4.0.0 (v4 in this case, but it will vary based on the package)

> Again, all I'm asking at the moment is to do the npm deprecate <package-spec> <message> step

I'm with you, I can do it over the weekend in case if nobody is willing to do it before

**Comment by jsumners** [comment]
> I'm with you, I can do it over the weekend in case if nobody is willing to do it before

I am unlikely to have time today. I have a busy schedule.

**Comment by Eomm** [comment]
Done

I don't have power for one module, @fastify/leads could you run this command?

```
npm deprecate fastify-error@1.2.0 "Please use @fastify/error@2.0.0 instead" --otp <OTP>
```


<details>
  <summary>Script</summary>

```
npm deprecate fastify-autoload@3.13.0 "Please use @fastify/autoload@4.0.0 instead" --otp
npm deprecate fastify-accepts@2.3.0 "Please use @fastify/accepts@3.0.0 instead" --otp
npm deprecate fastify-accepts-serializer@3.3.0 "Please use @fastify/accepts-serializer@4.0.0 instead" --otp
npm deprecate fastify-auth@1.2.0 "Please use @fastify/auth@2.0.0 instead" --otp
npm deprecate fastify-basic-auth@2.3.0 "Please use @fastify/basic-auth@3.0.0 instead" --otp
npm deprecate fastify-bearer-auth@6.3.0 "Please use @fastify/bearer-auth@7.0.0 instead" --otp
npm deprecate fastify-caching@6.3.0 "Please use @fastify/caching@7.0.0 instead" --otp
npm deprecate fastify-circuit-breaker@1.2.0 "Please use @fastify/circuit-breaker@2.0.0 instead" --otp
npm deprecate fastify-compress@4.1.0 "Please use @fastify/compress@5.0.0 instead" --otp
npm deprecate fastify-cookie@5.7.0 "Please use @fastify/cookie@6.0.0 instead" --otp
npm deprecate fastify-cors@6.1.0 "Please use @fastify/cors@7.0.0 instead" --otp
npm deprecate fastify-diagnostics-channel@1.1.0 "Please use @fastify/diagnostics-channel@2.0.0 instead" --otp
npm deprecate fastify-elasticsearch@1.2.0 "Please use @fastify/elasticsearch@2.0.0 instead" --otp
npm deprecate fastify-env@2.2.0 "Please use @fastify/env@3.0.0 instead" --otp
npm deprecate fastify-etag@2.2.0 "Please use @fastify/etag@3.0.0 instead" --otp
npm deprecate fastify-express@0.4.0 "Please use @fastify/express@1.0.0 instead" --otp
npm deprecate fastify-flash@3.1.0 "Please use @fastify/flash@4.0.0 instead" --otp
npm deprecate fastify-formbody@5.3.0 "Please use @fastify/formbody@6.0.0 instead" --otp
npm deprecate fastify-funky@1.1.0 "Please use @fastify/funky@2.0.0 instead" --otp
npm deprecate fastify-helmet@7.1.0 "Please use @fastify/helmet@8.0.0 instead" --otp
npm deprecate fastify-hotwire@0.3.0 "Please use @fastify/hotwire@1.0.0 instead" --otp
npm deprecate fastify-http-proxy@6.3.0 "Please use @fastify/http-proxy@7.0.0 instead" --otp
npm deprecate fastify-jwt@4.2.0 "Please use @fastify/jwt@5.0.0 instead" --otp
npm deprecate fastify-kafka@0.3.0 "Please use @fastify/kafka@1.0.0 instead" --otp
npm deprecate fastify-leveldb@3.2.0 "Please use @fastify/leveldb@4.0.0 instead" --otp
npm deprecate fastify-mongodb@4.2.0 "Please use @fastify/mongodb@5.0.0 instead" --otp
npm deprecate fastify-multipart@5.4.0 "Please use @fastify/multipart@6.0.0 instead" --otp
npm deprecate fastify-mysql@2.2.0 "Please use @fastify/mysql@3.0.0 instead" --otp
npm deprecate fastify-nextjs@7.3.0 "Please use @fastify/nextjs@8.0.0 instead" --otp
npm deprecate fastify-oauth2@4.6.0 "Please use @fastify/oauth2@5.0.0 instead" --otp
npm deprecate fastify-passport@0.5.0 "Please use @fastify/passport@1.0.0 instead" --otp
npm deprecate fastify-postgres@3.7.0 "Please use @fastify/postgres@4.0.0 instead" --otp
npm deprecate fastify-rate-limit@5.9.0 "Please use @fastify/rate-limit@6.0.0 instead" --otp
npm deprecate fastify-redis@4.4.0 "Please use @fastify/redis@5.0.0 instead" --otp
npm deprecate fastify-reply-from@6.7.0 "Please use @fastify/reply-from@7.0.0 instead" --otp
npm deprecate fastify-request-context@2.3.0 "Please use @fastify/request-context@3.0.0 instead" --otp
npm deprecate fastify-response-validation@0.3.0 "Please use @fastify/response-validation@1.0.0 instead" --otp
npm deprecate fastify-routes@3.2.0 "Please use @fastify/routes@4.0.0 instead" --otp
npm deprecate fastify-routes-stats@2.1.0 "Please use @fastify/routes-stats@3.0.0 instead" --otp
npm deprecate fastify-secure-session@3.2.0 "Please use @fastify/secure-session@4.0.0 instead" --otp
npm deprecate fastify-sensible@3.2.0 "Please use @fastify/sensible@4.0.0 instead" --otp
npm deprecate fastify-soap-client@0.3.0 "Please use @fastify/soap-client@1.0.0 instead" --otp
npm deprecate fastify-static@4.7.0 "Please use @fastify/static@5.0.0 instead" --otp
npm deprecate fastify-swagger@5.2.0 "Please use @fastify/swagger@6.0.0 instead" --otp
npm deprecate fastify-url-data@3.1.0 "Please use @fastify/url-data@4.0.0 instead" --otp
npm deprecate fastify-websocket@4.3.0 "Please use @fastify/websocket@5.0.0 instead" --otp
npm deprecate fastify-zipkin@2.1.0 "Please use @fastify/zipkin@3.0.0 instead" --otp
```
</details>

**Comment by jsumners** [comment]
Done.
